sqlx-postgres/src/bind_iter.rs

```diff
@@ -0,0 +1,114 @@
+use sqlx_core::{
+    database::Database,
+    encode::{Encode, IsNull},
+    types::Type,
+};
+
+use crate::{type_info::PgType, PgArgumentBuffer, PgHasArrayType, PgTypeInfo, Postgres};
+
+pub struct PgBindIter<I>(I);
+
+impl<I> PgBindIter<I> {
+    pub fn new(inner: I) -> Self {
+        Self(inner)
+    }
+}
+
+impl<I> From<I> for PgBindIter<I> {
+    fn from(inner: I) -> Self {
+        Self::new(inner)
+    }
+}
+
+impl<T, I> Type<Postgres> for PgBindIter<I>
+where
+    T: Type<Postgres> + PgHasArrayType,
+    I: Iterator<Item = T>,
+{
+    fn type_info() -> <Postgres as Database>::TypeInfo {
+        T::array_type_info()
+    }
+    fn compatible(ty: &PgTypeInfo) -> bool {
+        T::array_compatible(ty)
+    }
+}
+
+impl<'q, T, I> PgBindIter<I>
+where
+    I: Iterator<Item = T>,
+    T: Type<Postgres> + Encode<'q, Postgres>,
+{
+    fn encode_inner(
+        // need ownership to iterate
+        mut iter: I,
+        buf: &mut PgArgumentBuffer,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let first = iter.next();
+        let type_info = first
+            .as_ref()
+            .and_then(Encode::produces)
+            .unwrap_or_else(T::type_info);
+
+        buf.extend(&1_i32.to_be_bytes()); // number of dimensions
+        buf.extend(&0_i32.to_be_bytes()); // flags
+
+        match type_info.0 {
+            PgType::DeclareWithName(name) => buf.patch_type_by_name(&name),
+            PgType::DeclareArrayOf(array) => buf.patch_array_type(array),
+
+            ty => {
+                buf.extend(&ty.oid().0.to_be_bytes());
+            }
+        }
+
+        let len_start = buf.len();
+        buf.extend(0_i32.to_be_bytes()); // len (unknown so far)
+        buf.extend(1_i32.to_be_bytes()); // lower bound
+
+        match first {
+            Some(first) => buf.encode(first)?,
+            None => return Ok(IsNull::No),
+        }
+
+        let mut count = 1_i32;
+        const MAX: usize = i32::MAX as usize;
```

Comment:
This looks like an oversight. We've already written 1 item at this point.

```suggestion
        const MAX: usize = i32::MAX as usize - 1;
```

Reply:
Good catch. I missed this when removing the peeking iterator.